app/controllers/CheckYourAnswersController.scala

```diff
@@ -151,81 +156,143 @@ class CheckYourAnswersController @Inject()(
-                auditService.audit(ReturnsAuditModel.build(
-                  vatReturnRequest, None, SubmissionResult.Failure, None, None, request
-                ))
-                Redirect(routes.JourneyRecoveryController.onPageLoad()).toFuture
-            }
-          case Invalid(errors) =>
-            val errorList = errors.toChain.toList
-            val errorMessages = errorList.map(_.errorMessage).mkString("\n")
-            logger.error(s"Unable to create a VAT return request from user answers: $errorMessages")
-
-            Redirect(routes.JourneyRecoveryController.onPageLoad()).toFuture
-        }
+
+      val redirectToFirstError = getRedirect(validate(period), period).headOption
+
+      (redirectToFirstError, incompletePromptShown) match {
+        case (Some(c), true) => Future.successful(Redirect(c))
+        case (Some(_), false) => Future.successful(Redirect(routes.CheckYourAnswersController.onPageLoad(period)))
+        case _ =>
+          val validatedVatReturnRequest =
+            vatReturnService.fromUserAnswers(request.userAnswers, request.vrn, period, request.registration)
+
+          val validatedCorrectionRequest = request.userAnswers.get(CorrectPreviousReturnPage).map(_ =>
+            correctionService.fromUserAnswers(request.userAnswers, request.vrn, period, request.registration.commencementDate))
+
+          (validatedVatReturnRequest, validatedCorrectionRequest) match {
+            case (Valid(vatReturnRequest), Some(Valid(correctionRequest))) =>
+              submitReturn(vatReturnRequest, Option(correctionRequest), period)
+            case (Valid(vatReturnRequest), None) =>
+              submitReturn(vatReturnRequest, None, period)
+            case (Invalid(vatReturnErrors), Some(Invalid(correctionErrors))) =>
+              val errors = vatReturnErrors ++ correctionErrors
+              val errorList = errors.toChain.toList
+              val errorMessages = errorList.map(_.errorMessage).mkString("\n")
+              logger.error(s"Unable to create a vat return and correction request from user answers: $errorMessages")
+              Redirect(routes.JourneyRecoveryController.onPageLoad()).toFuture
+            case (Invalid(errors), _) =>
+              val errorList = errors.toChain.toList
+              val errorMessages = errorList.map(_.errorMessage).mkString("\n")
+              logger.error(s"Unable to create a vat return request from user answers: $errorMessages")
+              Redirect(routes.JourneyRecoveryController.onPageLoad()).toFuture
+            case (_, Some(Invalid(errors))) =>
+              val errorList = errors.toChain.toList
+              val errorMessages = errorList.map(_.errorMessage).mkString("\n")
+              logger.error(s"Unable to create a Corrections request from user answers: $errorMessages")
+              Redirect(routes.JourneyRecoveryController.onPageLoad()).toFuture
+          }
       }
+  }
+
+  //noinspection ScalaStyle
```

Comment:
Could we remove this comment and fix any subsequent warnings

Reply:
It was a cyclomatic complexity warning. As there are so many case clauses I can't see a simple way of fixing this warning without complicating things. I'm open to suggestions!